src/config.cc

```diff
@@ -5,200 +5,198 @@
-						CELERITY_WARN("CELERITY_DEVICES contains {} device indices, but only {} worker processes were spawned on this host", values.size() - 1,
-						    m_host_cfg.node_count);
-					}
-
-					const auto pid_parsed = parse_uint(values[0].c_str());
-					const auto did_parsed = parse_uint(values[m_host_cfg.local_rank + 1].c_str());
-					if(!pid_parsed.first || !did_parsed.first) {
-						CELERITY_WARN("CELERITY_DEVICES contains invalid value(s) - will be ignored");
-					} else {
-						m_device_cfg = device_config{pid_parsed.second, did_parsed.second};
-					}
-				}
-			}
-		}
+			// --------------------------------- CELERITY_DEVICES ---------------------------------
 
-		// ----------------------------- CELERITY_PROFILE_KERNEL ------------------------------
-		{
-			const auto result = get_env("CELERITY_PROFILE_OCL");
-			if(result.first) {
-				CELERITY_WARN("CELERITY_PROFILE_OCL has been renamed to CELERITY_PROFILE_KERNEL with Celerity 0.3.0.");
-				m_enable_device_profiling = result.second == "1";
+			const auto has_devs = parsed_and_validated_envs.get(env_devs);
+			if(has_devs) {
+				m_devices_list = *has_devs;
+				const auto pid_parsed = (*has_devs)[0];
+				const auto did_parsed = (*has_devs)[m_host_cfg.local_rank + 1];
+				m_device_cfg = device_config{pid_parsed, did_parsed};
 			}
-		}
 
-		{
-			const auto result = get_env("CELERITY_PROFILE_KERNEL");
-			if(result.first) { m_enable_device_profiling = result.second == "1"; }
-		}
+			// ----------------------------- CELERITY_PROFILE_KERNEL ------------------------------
 
-		// -------------------------------- CELERITY_FORCE_WG ---------------------------------
+			const auto has_profile_kernel = parsed_and_validated_envs.get(env_profile_kernel);
+			if(has_profile_kernel) { m_enable_device_profiling = *has_profile_kernel; }
 
-		{
-			const auto result = get_env("CELERITY_FORCE_WG");
-			if(result.first) { CELERITY_WARN("Support for CELERITY_FORCE_WG has been removed with Celerity 0.3.0."); }
-		}
+			// -------------------------------- CELERITY_DRY_RUN_NODES ---------------------------------
 
-		// -------------------------------- CELERITY_DRY_RUN_NODES ---------------------------------
-		{
-			const auto [is_set, value] = get_env("CELERITY_DRY_RUN_NODES");
-			if(is_set) {
-				const auto [is_valid, num_nodes] = parse_uint(value.c_str());
-				if(!is_valid) { CELERITY_WARN("CELERITY_DRY_RUN_NODES contains invalid value - will be ignored"); }
-				m_dry_run_nodes = num_nodes;
+			const auto has_dry_run_nodes = parsed_and_validated_envs.get(env_dry_run_nodes);
+			if(has_dry_run_nodes) { m_dry_run_nodes = *has_dry_run_nodes; }
+
+		} else {
+			for(const auto& warn : parsed_and_validated_envs.errors()) {
+				CELERITY_WARN(warn.what());
+			}
+			for(const auto& err : parsed_and_validated_envs.warnings()) {
+				CELERITY_ERROR(err.what());
 			}
+		throw std::runtime_error("Please make sure to use the environment variables correctly.");
```

Comment:
Wrong indentation.
Also I'm not sure about this error message. I feel like telling the user to use the environment variables correctly is neither useful nor necessary. A much simpler `"Failed to parse/validate environment variables"` would probably be suffcient without sounding patronizing.